packages/@sanity/client/sanityClient.d.ts

```diff
@@ -0,0 +1,1784 @@
+   */
+  namespace?: string
+
+  /**
+   * @deprecated Don't use
+   */
+  useProjectHostname?: boolean
+
+  /**
+   * @deprecated Don't use
+   */
+  requester?: GetItRequester
+}
+
+type ProjectlessClientConfig = BaseClientConfig & {
+  useProjectHostname: false
+  projectId?: string
+  dataset?: string
+}
+
+type ProjectClientConfig = {
+  projectId: string
+  dataset?: string
+} & BaseClientConfig
+
+type ClientConfig = ProjectClientConfig | ProjectlessClientConfig
+
+interface RequestOptions {
+  timeout?: number
+  token?: string
+}
+
+type BaseMutationOptions = RequestOptions & {
+  visibility?: 'sync' | 'async' | 'defer'
+  returnDocuments?: boolean
+  returnFirst?: boolean
+}
+
+type MutationEvent = {
+  type: 'mutation'
+  documentId: string
+  eventId: string
+  identity: string
+  mutations: Mutation[]
+  previousRev?: string
+  resultRev?: string
+  result?: SanityDocument
+  previous?: SanityDocument | null
+  timestamp: string
+  transactionId: string
+  transition: 'update' | 'appear' | 'disappear'
+}
+
+type ChannelErrorEvent = {
+  type: 'channelError'
+  message: string
+}
+
+type DisconnectEvent = {
+  type: 'disconnect'
+  reason: string
+}
+
+type ReconnectEvent = {
+  type: 'reconnect'
+}
+
+type WelcomeEvent = {
+  type: 'welcome'
+}
+
+type ListenEvent =
+  | MutationEvent
+  | ChannelErrorEvent
+  | DisconnectEvent
+  | ReconnectEvent
+  | WelcomeEvent
+
+interface ListenOptions {
+  includeResult?: boolean
+  includePreviousRevision?: boolean
+  visibility?: 'sync' | 'async' | 'query'
+  events?: ListenEventName[]
+}
+
+type PreviousNextListenOptions = ListenOptions & {
+  includeResult: true
+  includePreviousRevision: true
+}
+
+type PreviousListenOptions = ListenOptions & {
+  includePreviousRevision: true
+  includeResult: false
+}
+
+type NextListenOptions = ListenOptions & {
+  includePreviousRevision: false
+  includeResult: true
+}
+
+type ResultlessListenOptions = ListenOptions & {
+  includeResult: false
+  includePreviousRevision: false
+}
+
+type FilteredResponseQueryOptions = RequestOptions & {
+  filterResponse: true
+}
+
+type UnfilteredReponseQueryOptions = RequestOptions & {
+  filterResponse: false
+}
+
+type QueryOptions = FilteredResponseQueryOptions | UnfilteredReponseQueryOptions
+
+type FirstDocumentMutationOptions = BaseMutationOptions & {
+  returnFirst?: true
+  returnDocuments?: true
+}
+
+type FirstDocumentIdMutationOptions = BaseMutationOptions & {
+  returnFirst?: true
+  returnDocuments: false
+}
+
+type AllDocumentsMutationOptions = BaseMutationOptions & {
+  returnFirst: false
+  returnDocuments?: true
+}
+
+type AllDocumentIdsMutationOptions = BaseMutationOptions & {
+  returnFirst: false
+  returnDocuments: false
+}
+
+type MutationOptions =
+  | FirstDocumentMutationOptions
+  | FirstDocumentIdMutationOptions
+  | AllDocumentsMutationOptions
+  | AllDocumentIdsMutationOptions
+
+interface RawQueryResponse {
+  q: string
+  ms: number
+  result: any
+}
+
+interface SanityDocument {
```

Comment:
`interface SanityDocument<T = {}>` + `[P in keyof T]: T[P]` alongside `[key: string]: any`.

(I may have gotten the syntax wrong, on a quick pass to set down all notes)

Reply:
Yeah, the syntax would be:

```
type SanityDocument<T extends object = {}> = {
  [P in keyof T]: T[P]
} & {
  _id: string
  _rev: string
  _type: string
  _createdAt: string
  _updatedAt: string
  [key: string]: any
}
```

Reply:
Thanks so much for this - I'm still fairly new to typescript, and this was very useful. I've now implemented this throughout.